examples/vision/siamese.py

```diff
@@ -0,0 +1,232 @@
+x_test = x_test.astype('float32')
+
+"""
+## Create pairs of images
+
+We want our model to be able to differentiate 0 from 1,2,3, .., 9 to do so we
+should teach it how each digit differs from each of the other digit, for this
+purpose we would select N random images from class A (lets say class of
+digit 0) and would pair it with N random images from class B (lets say class of
+digit 1), and would repeat this process untill class of digit 9. In this way we
+successfuly pair 0 with each other digit, now we would do the same for class of
+1,2,3 upto that of 9.
+"""
+
+def make_pairs(x, y):
+    num_classes = max(y) + 1
+    digit_indices = [np.where(y == i)[0] for i in range(num_classes)]
+
+    pairs = []
+    labels = []
+
+    for idx1 in range(len(x)):
+        # add a matching example
+        x1 = x[idx1]
+        label1 = y[idx1]
+        idx2 = random.choice(digit_indices[label1])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [1]
+    
+        # add a not matching example
+        label2 = random.randint(0, num_classes-1)
+        while label2 == label1:
+            label2 = random.randint(0, num_classes-1)
+
+        idx2 = random.choice(digit_indices[label2])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [0]
+
+    return np.array(pairs), np.array(labels).astype('float32')
+
+pairs_train, labels_train = make_pairs(x_train, y_train)
+pairs_test, labels_test = make_pairs(x_test, y_test)
+
+"""
+## Create dataset
+"""
+
+# 0 index of pairs_train contain one image of pair
+# and at index 1 the other image of pair is present
+# we can split entire vector of pairs into two havles,
+# later we can generate tf.data.Dataset using these halves
+# we will do the same for labels
+
+x_train_1 = pairs_train[:,0]
+x_train_2 = pairs_train[:,1]
+# x1.shape = (120000, 28, 28)
+
+x_test_1, x_test_2 = pairs_test[:,0],pairs_test[:,1]
+
+train_pair = tf.data.Dataset.from_tensor_slices((x_train_1, x_train_2))
+train_label = tf.data.Dataset.from_tensor_slices(labels_train)
+train_ds = tf.data.Dataset.zip((train_pair, train_label)).batch(16)
+
+test_pair = tf.data.Dataset.from_tensor_slices((x_test_1, x_test_2))
+test_label = tf.data.Dataset.from_tensor_slices(labels_test)
+test_ds = tf.data.Dataset.zip((test_pair, test_label)).batch(16)
+
+"""
+## Visualize
+"""
+
+# images to show must be integral multiple of num_col
+# if not then then they would be made so
+def visualize(dataset, to_show=6, num_col=3, predictions=None, test=False):
+
+  # handle case when user input more columns then images to show
+  num_row = to_show//num_col if to_show//num_col != 0 else 1 
+
+  # num_col must be integral multiple of to_show 
+  to_show = to_show if to_show/num_col == 0 else num_row*num_col
+
+  # plot images
+  fig, axes = plt.subplots(num_row, num_col, figsize=(5,5))
+  for images, labels in dataset.take(1):
+    for i in range(to_show):
+
+        # if row is one, axes array is one dimentional
```

Comment:
Suggestion:

- Fix typo
- Refactor

```suggestion
        # If the number of rows is 1, the axes array is one-dimensional
```